### PR TITLE
HYP-24: controller for revocation

### PR DESCRIPTION
**Pull request by n3op2**
# What

Adding ability to revoke an issued certificate by an entity of "regulator" in this instance it's -> "Charlie" 5FLSigC9HGRKVhB9FiEo4Y3koPsNmBmLJbpXg2mp1hXcS59Y

Changes
- README update to contain some info about the roles since they are less likelyt to change
- Added some TODOs to readme for project wise "nice to haves" that also reflects a backlog items
- moved Zod outside of db.ts
- some changes to process flows and payloads
- new controllers to handle revocation
- update to dockerc-compsoe 3 persona

> TODO
```
**{"level":40,"time":1702457576712,"pid":3365,"hostname":"fv-az698-227","name":"dscp-hyproof-api","msg":"Error thrown in handler: {\"Message\":\"multipart: NextPart: EOF\",\"Code\":0,\"Type\":\"error\"}\n"}
{"level":40,"time":1702457576717,"pid":3365,"hostname":"fv-az698-227","name":"dscp-hyproof-api","msg":"Handled Validation Error for /v1/certificate/c1b09077-b27c-4696-b765-6c488f1ca540/revocation: {\".reason\":{\"message\":\"'reason' is required\"}}"}
        1) should revoke an issued certificate with a reason as an attachment**
```

Screenshot 2023-12-10 at 19 24 15
Screenshot 2023-12-10 at 19 17 07
Screenshot 2023-12-10 at 19 15 02
Screenshot 2023-12-10 at 19 13 47

### CODE REVIEW

File: src/controllers/v1/certificate/index.ts [hunk 1]
```diff
@@ -302,8 +302,9 @@ export class CertificateController extends Controller {
 
   /**
    * Update a certificate on-chain to include
-   * @summary Create a new demandA on-chain
-   * @param demandAId The demandA's identifier
+   * @summary updates initiated certitificate status to issued on chain
```
Comment (rmlearney-digicatapult) [inline src/controllers/v1/certificate/index.ts:305]: I think @mattdean-digicatapult fixed this in his previous PR. You may want to rebase

File: src/lib/db/migrations/20230310111029_initial.ts [hunk 2]
```diff
@@ -40,7 +40,9 @@ export async function up(knex: Knex): Promise<void> {
     def.integer('original_token_id').defaultTo(null)
     def.datetime('created_at').notNullable().defaultTo(now())
     def.datetime('updated_at').notNullable().defaultTo(now())
+    def.uuid('reason').notNullable()
```
Comment (mattdean-digicatapult) [inline src/lib/db/migrations/20230310111029_initial.ts:43]: this will need to be nullable as we only have a revocation reason if we have revoked a cert

File: src/lib/indexer/eventProcessor.ts [hunk 3]
```diff
@@ -38,13 +38,13 @@ const parseIntegerOrThrow = (value: string): number => {
   return result
 }
 
-/* INFO uncomment if we decided to use attachments
 const attachmentPayload = (map: Map<string, string>, key: string): AttachmentRecord => ({
   type: 'insert',
   id: UUIDv4(),
   ipfs_hash: getOrError(map, key),
+  filename: 'reason',
+  size: 0,
```
Comment (mattdean-digicatapult) [inline src/lib/indexer/eventProcessor.ts:46]: these should be `null` I think. We don't yet know the filename or size and that's ok. Note these come from the attachment upload on the other side and are stored in ipfs

File: src/lib/db/migrations/20230310111029_initial.ts (hunk 2 above)
Comment (n3op2) [reply src/lib/db/migrations/20230310111029_initial.ts]: yea, this has been updated along with some other little changes on friday, did not got to the e2e

File: src/lib/indexer/eventProcessor.ts (hunk 3 above)
Comment (n3op2) [reply src/lib/indexer/eventProcessor.ts:46]: yes, this could be it actually!

File: README.md [hunk 4]
```diff
@@ -107,19 +107,54 @@ npm run flows
...
+running with docker-compose -f docker-compose-3-personal.yml logs -f will render all logs, also can be parsed by `| grep`.
+```
+docker-compose -f docker-compose-3-personal.yml logs -f | grep regulator
+```
 
+> single persona
 Run `docker composel up -d` to start the required dependencies to fully demo `dscp-hyproof-api`.
 
+> multiple persona
+Run `docker-compose -f docker-compose-3-personal.yml up` to start the required dependencies to fully demo `dscp-hyproof-api`.
```
Comment (rmlearney-digicatapult) [inline README.md:125]: docker-compose-3-persona.yml (personal -> persona)

File: README.md [hunk 5]
```diff
@@ -107,19 +107,54 @@ npm run flows
...
-...update once clear
+# Identities / System wide roles
+The regulator identity must be includd as part of a draft certificate which will be carried away until certificate is revoked or further features implemented.
+
+> Energy Owner (Emma) must submit a commitment before issuing a cert
+- [API] - localhost:8000/swagger
+- [address] - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY 
+- [alias] - //Alice
+
+> Hydrogen Owner (Haidi) commitement is generated along by consumption and production which is verified by energy owner Emma
```
Comment (rmlearney-digicatapult) [inline README.md:141]: Heidi the Hydrogen Producer generates a commitment from her production start and end time stamps and total energy consumption. This is verified by Emma the Energy Owner.

File: README.md [hunk 6]
```diff
@@ -107,19 +107,54 @@ npm run flows
...
 - dscp-identity-service (+ PostgreSQL)
+- ipfs
 - dscp-node
 
-### Identities
-...update once clear
+# Identities / System wide roles
+The regulator identity must be includd as part of a draft certificate which will be carried away until certificate is revoked or further features implemented.
+
+> Energy Owner (Emma) must submit a commitment before issuing a cert
```
Comment (rmlearney-digicatapult) [inline README.md:136]: What data does Emma commit to?

File: README.md [hunk 7]
```diff
@@ -107,19 +107,54 @@ npm run flows
...
+- [address] - 5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY 
+- [alias] - //Alice
+
+> Hydrogen Owner (Haidi) commitement is generated along by consumption and production which is verified by energy owner Emma
+- [API] - localhost:8010/swagger
+- [address] - 5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty
+- [alias] - //Bob
 
-### Using the hyproof API
+> Regulator (Reginald?) has the power to oversee the future, however, it has the ability to revoke an issued certificate but before that must provide a reason (ideally valid
```
Comment (rmlearney-digicatapult) [inline README.md:146]: Reginald the Regulator has the power to revoke certificates but must provide a documented reason which is attached to the certificate and stored in IPFS.

File: processFlows.dscp
Comment (rmlearney-digicatapult) [inline processFlows.dscp:1]: Submit process flow changes as separate PR from API controller?

File: processFlows.json
Comment (rmlearney-digicatapult) [inline processFlows.json:1]: Submit process flow changes as separate PR from API controller?

File: src/lib/db/types.ts
Comment (rmlearney-digicatapult) [inline src/lib/db/types.ts:1]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: src/lib/indexer/__tests__/eventProcessor.test.ts
Comment (rmlearney-digicatapult) [inline src/lib/indexer/__tests__/eventProcessor.test.ts:1]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: src/lib/indexer/eventProcessor.ts [hunk 8]
```diff
@@ -1,10 +1,10 @@
 import { v4 as UUIDv4 } from 'uuid'
 
 import { UUID } from '../../models/strings'
-import { TransactionRow } from '../db'
-import { ChangeSet, CertificateRecord } from './changeSet'
+import { TransactionRow } from '../../lib/db/types'
```
Comment (rmlearney-digicatapult) [inline src/lib/indexer/eventProcessor.ts:4]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: src/lib/payload.ts [hunk 9]
```diff
@@ -1,4 +1,5 @@
-import { CertificateRow } from './db'
+import { AttachmentRow, CertificateRow } from './db/types'
```
Comment (rmlearney-digicatapult) [inline src/lib/payload.ts:1]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: test/helpers/chainTest.ts [hunk 10]
```diff
@@ -5,7 +5,8 @@ import { container } from 'tsyringe'
 import createHttpServer from '../../src/server'
 import Indexer from '../../src/lib/indexer'
 
-import Database, { CertificateRow } from '../../src/lib/db'
+import { CertificateRow } from '../../src/lib/db/types'
```
Comment (rmlearney-digicatapult) [inline test/helpers/chainTest.ts:8]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: test/integration/offchain/certificate.test.ts [hunk 11]
```diff
@@ -6,7 +6,7 @@ import createHttpServer from '../../../src/server'
 import { put } from '../../helpers/routeHelper'
 
 import { cleanup, updateSeed } from '../../seeds/certificate'
-import { CertificateRow } from '../../../src/lib/db'
+import { CertificateRow } from '../../../src/lib/db/types'
```
Comment (rmlearney-digicatapult) [inline test/integration/offchain/certificate.test.ts:9]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: test/integration/onchain/certificate.test.ts [hunk 12]
```diff
@@ -17,13 +17,15 @@ import {
   regulatorAlias,
   regulatorAddress,
 } from '../../helpers/mock'
-import Database, { CertificateRow } from '../../../src/lib/db'
+import Database from '../../../src/lib/db'
+import { CertificateRow } from '../../../src/lib/db/types'
```
Comment (rmlearney-digicatapult) [inline test/integration/onchain/certificate.test.ts:21]: DB changes as separate PR from API controller, testing first with initiation and issuance?

File: processFlows.dscp [hunk 13]
```diff
@@ -27,7 +27,7 @@ token RevokedCert {
   hydrogen_owner: Role,
   energy_owner: Role,
   regulator: Role,
-  reason: File,
+  reason: File | None,
```
Comment (mattdean-digicatapult) [inline processFlows.dscp:30]: why `| None`

File: src/controllers/v1/certificate/index.ts [hunk 14]
```diff
@@ -367,4 +369,80 @@ export class CertificateController extends Controller {
...
+   * @summary returns transactions by certificate local id
+   * @example id "52907745-7672-470e-a803-a2f8feb52944"
+   */
+  @Response<ValidateError>(422, 'Validation Failed')
+  @Response<NotFound>(404, '<item> not found')
+  @Response<BadRequest>(400, 'ID must be supplied in UUID format')
+  @Get('{id}/revocation')
+  public async getRevocationTransactions(@Path() id: UUID): Promise<Certificate.ListTransactionResponse> {
+    if (!id) throw new BadRequest()
+    return await this.db.get('transaction', { local_id: id, api_type: 'certificate', transaction_type: 'issue_cert' })
```
Comment (mattdean-digicatapult) [inline src/controllers/v1/certificate/index.ts:408]: `transaction_type` needs to be `revoke_cert`

File: src/lib/indexer/eventProcessor.ts (hunk 8 above)
Comment (mattdean-digicatapult) [reply src/lib/indexer/eventProcessor.ts:4]: I think this is fine. The DB changes look pretty clean

File: src/lib/indexer/eventProcessor.ts [hunk 15]
```diff
@@ -112,6 +112,26 @@ const DefaultEventProcessors: EventProcessors = {
       certificates: new Map([[local_id, update]]),
     }
   },
+  revoke_cert: ({ version, inputs, outputs }) => {
+    if (version !== 1) throw new Error(`Incompatible version ${version} for issue_cert process`)
+
+    const { local_id } = inputs[0]
+    const { id: latest_token_id, ...cert } = outputs[0]
+
+    const attachment: AttachmentRecord = attachmentPayload(cert.metadata, 'reason')
```
Comment (mattdean-digicatapult) [inline src/lib/indexer/eventProcessor.ts:121]: The behaviour here looks incorrect. If it's our transaction we need to do something different. See https://github.com/digicatapult/dscp-matchmaker-api/blob/1ef781b85e1273de19f5347725e9ac5dc1191990/src/lib/indexer/eventProcessor.ts#L90 for an example

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
A few necessary changes, but getting close

**Review by mattdean-digicatapult — APPROVED** [review]
Approved, but I'll need another owner to review as I'm the one who's finished this off

**Review by dblane-digicatapult — APPROVED** [review]
